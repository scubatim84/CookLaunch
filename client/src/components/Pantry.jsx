import React, {useEffect, useState} from 'react';
import {Redirect} from 'react-router-dom';
import _ from 'lodash';
import {
  addIngredientToPantry,
  deleteIngredientFromPantry,
  updateIngredientInPantry,
} from '../actions/pantryActions';
import IngredientAdd from './Ingredients/IngredientAdd';
import IngredientItem from './Ingredients/IngredientItem';
import {useStylesMain} from '../Styles';
import {themeMain} from '../Theme';
import {Card, Container, Grid, List} from '@material-ui/core';
import CardTitle from './CardTitle';
import Loader from './Loader';

function Pantry(props) {
  const classes = useStylesMain(themeMain);

  const [pantry, setPantry] = useState(null);

  useEffect(() => {
    if (props.pantry && props.pantry.length > 0) {
      let rawPantryList = props.pantry;
      let sortedPantry = rawPantryList.sort((a, b) => {
        return a.name.localeCompare(b.name);
      });

      setPantry({data: sortedPantry});
    }
  }, [props.pantry]);

  const handleDelete = async (ingredientId) => {
    const response = await deleteIngredientFromPantry(ingredientId);

    if (response.status === 204) {
      // Update user payload to re-render pantry
      await props.getUserPayload();
    } else {
      // If request failed, return error message to child component
      return response.data;
    }
  };

  const handleUpdateIngredient = async (updateIngredient) => {
    const response = await updateIngredientInPantry(updateIngredient);

    if (response.status === 204) {
      // Update user payload to re-render pantry
      await props.getUserPayload();
    } else {
      // If request failed, return error message to child component
      return response.data;
    }
  };

  const handleAddIngredient = async (addIngredient) => {
    const response = await addIngredientToPantry(addIngredient);

    if (response.status === 201) {
      // Update user payload to re-render pantry
      await props.getUserPayload();
    } else {
      return response.data;
    }
  };

  if (!props.isLoggedIn) {
    return <Redirect to='/login' />;
  } else if (!pantry) {
    return <Loader />;
  } else {
    return (
      <Container component='main' maxWidth='xs'>
        <Grid container spacing={1} className={classes.pageMarginBottom}>
          <Grid item xs={12}>
            <Card>
              <div className={classes.paper}>
                <Grid container spacing={2}>
                  <Grid item xs={12} align='center'>
                    <CardTitle title={`${props.firstName}'s Pantry`} />
                  </Grid>
                  <Grid item xs={12}>
                    <List className={classes.list}>
                      {pantry &&
                        pantry.data.map((ingredient) => {
                          const formatName = _.startCase(
                            _.toLower(ingredient.name)
                          );
                          const formatQuantityType = _.startCase(
                            _.toLower(ingredient.quantityType)
                          );

                          return (
                            <Grid
                              item
                              xs={12}
                              key={ingredient.name + ingredient.dateLastChanged}
                            >
                              <IngredientItem
                                key={
                                  ingredient.name + ingredient.dateLastChanged
                                }
                                id={ingredient._id}
                                name={formatName}
                                quantity={ingredient.quantity}
                                quantityType={formatQuantityType}
                                handleDelete={handleDelete}
                                handleUpdateIngredient={handleUpdateIngredient}
                              />
                            </Grid>
                          );
                        })}
                    </List>
                  </Grid>
                </Grid>
              </div>
            </Card>
          </Grid>
          <Grid item xs={12}>
            <IngredientAdd
              key={props.pantry}
              name='Pantry'
              pantry={props.pantry}
              ingredients={props.ingredients}
              handleAddIngredient={handleAddIngredient}
              handleUpdateIngredient={handleUpdateIngredient}
            />
          </Grid>
        </Grid>
      </Container>
    );
  }
}

export default Pantry;
