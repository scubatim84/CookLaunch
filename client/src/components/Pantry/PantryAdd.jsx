import React, {useState} from 'react';
import isEmpty from 'is-empty';
import {ingredientQuantityTypes} from '../../actions/types';
import {addIngredientToPantry} from '../../actions/pantryActions';
import FormSubmitMessage from '../FormSubmitMessage';
import {useStylesForm} from '../../Styles';
import {themeMain} from '../../Theme';
import {Button, Card, Grid, TextField, Typography} from '@material-ui/core';
import Autocomplete from '@material-ui/lab/Autocomplete';

function PantryAdd(props) {
  const classes = useStylesForm(themeMain);

  const [addIngredient, setAddIngredient] = useState({
    name: '',
    quantity: '',
    quantityType: '',
  });
  const [error, setError] = useState({
    errorMessage: '',
  });

  const handleAutocompleteName = async (event, ingredient) => {
    if (!isEmpty(ingredient)) {
      setAddIngredient((prevValue) => {
        return {
          ...prevValue,
          name: ingredient.name,
        };
      });
    }
  };

  const handleAutocompleteQuantityType = async (event, ingredientType) => {
    if (!isEmpty(ingredientType)) {
      setAddIngredient((prevValue) => {
        return {
          ...prevValue,
          quantityType: ingredientType,
        };
      });
    }
  };

  const handleChange = async (e) => {
    const {name, value} = e.target;

    setAddIngredient((prevValue) => {
      return {
        ...prevValue,
        [name]: value,
      };
    });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    const requestResponse = await addIngredientToPantry(addIngredient);

    if (requestResponse.status === 201) {
      // If adding ingredient is successful, clear form
      setAddIngredient({
        name: '',
        quantity: '',
        quantityType: '',
      });

      // If adding ingredient is successful, clear old errors
      setError({
        errorMessage: '',
      });

      // Update user payload to re-render pantry
      await props.getUserPayload();
    } else {
      setError({
        errorMessage: requestResponse.data,
      });
    }
  };

  return (
    <Card className={classes.root}>
      <div className={classes.paper}>
        <form noValidate>
          <Grid container spacing={3} className={classes.root}>
            <Grid item xs={12} align='center'>
              <Typography component='h1' variant='h5'>
                Add Ingredients To Pantry
              </Typography>
            </Grid>
            <Grid item xs={12} sm={5}>
              <Autocomplete
                id='ingredients'
                options={props.ingredients}
                onChange={handleAutocompleteName}
                getOptionLabel={(option) => option.name}
                renderInput={(params) => (
                  <TextField
                    {...params}
                    required
                    label='Ingredient Name'
                    variant='outlined'
                    id='name'
                    name='name'
                  />
                )}
              />
            </Grid>
            <Grid item xs={12} sm={2}>
              <TextField
                onChange={handleChange}
                variant='outlined'
                required
                value={addIngredient.quantity}
                label='Quantity'
                id='quantity'
                name='quantity'
              />
            </Grid>
            <Grid item xs={12} sm={3}>
              <Autocomplete
                id='ingredientQuantityTypes'
                options={ingredientQuantityTypes}
                onChange={handleAutocompleteQuantityType}
                getOptionLabel={(option) => option}
                renderInput={(params) => (
                  <TextField
                    {...params}
                    required
                    label='Quantity Type'
                    variant='outlined'
                    id='quantityType'
                    name='quantityType'
                  />
                )}
              />
            </Grid>
            <Grid item xs={12} sm={2}>
              <Button
                onClick={handleSubmit}
                fullWidth
                type='submit'
                variant='contained'
                color='primary'
                className={classes.submit}
              >
                Add
              </Button>
            </Grid>
            <Grid item xs={12}>
              {!isEmpty(error.errorMessage) && (
                <FormSubmitMessage submitMessage={error.errorMessage} />
              )}
            </Grid>
          </Grid>
        </form>
      </div>
    </Card>
  );
}

export default PantryAdd;
