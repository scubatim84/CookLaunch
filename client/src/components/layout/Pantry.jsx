import React, {useState, useEffect} from 'react';
import {Redirect} from 'react-router-dom';
import _ from 'lodash';
import isEmpty from 'is-empty';
import {REQUEST_SUCCESS} from '../../actions/types';
import {getUserData} from '../../actions/authActions';
import {
  addIngredientToPantry,
  getPantry,
  deleteIngredientFromPantry,
} from '../../actions/pantryActions';
import {useStylesForm} from '../../Styles';
import {themeMain} from '../../Theme';

import PantryItem from '../layout/PantryItem';
import FormSubmitMessage from '../layout/FormSubmitMessage';

import {
  Card,
  Container,
  CssBaseline,
  Grid,
  Typography,
} from '@material-ui/core';

function Pantry(props) {
  const classes = useStylesForm(themeMain);

  const [isLoggedIn, setLoggedIn] = useState(props.isLoggedIn);
  const [user, setUser] = useState({
    email: '',
    firstName: '',
    lastName: '',
  });
  const [error, setError] = useState({
    errorMessage: '',
  });
  const [pantry, setPantry] = useState({data: [{name: 'None'}]});
  const [addIngredient, setAddIngredient] = useState({
    name: '',
    quantity: '',
    quantityType: '',
  });

  const getPantryData = async () => {
    const response = await getPantry(user.email);

    if (response.authResponseType === REQUEST_SUCCESS) {
      setPantry({data: response.authResponsePayload});
    }
  };

  useEffect(() => {
    getPantryData(user.email);
  }, [user]);

  useEffect(() => {
    setLoggedIn(props.isLoggedIn);
  }, [props.isLoggedIn]);

  useEffect(() => {
    if (isLoggedIn) {
      const getUserPayload = async () => {
        const data = await getUserData();
        const userPayload = await data.payload;

        setUser({
          email: userPayload.email,
          firstName: userPayload.firstName,
          lastName: userPayload.lastName,
        });

        setPantry({
          data: userPayload.pantry,
        });
      };

      getUserPayload();
    }
  }, [isLoggedIn]);

  const handleChange = async (e) => {
    const {name, value} = e.target;

    setAddIngredient((prevValue) => {
      return {
        ...prevValue,
        [name]: value,
      };
    });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    const requestResponse = await addIngredientToPantry(addIngredient);

    if (requestResponse.authResponseType === REQUEST_SUCCESS) {
      // If adding ingredient is successful, clear form
      setAddIngredient({
        name: '',
        createdBy: user.email,
      });

      // If adding ingredient is successful, clear old errors
      setError({
        errorMessage: '',
      });

      // Update ingredient list
      await getPantryData();
    } else {
      setError({
        errorMessage: requestResponse.authResponsePayload,
      });
    }
  };

  const handleDelete = async (userEmail, ingredientId) => {
    await deleteIngredientFromPantry(userEmail, ingredientId);

    // Update pantry data to re-render
    await getPantryData();
  };

  return !isLoggedIn ? (
    <Redirect to='/login' />
  ) : (
    <Container component='main' maxWidth='sm'>
      <Card className={classes.root}>
        <CssBaseline />
        <div className={classes.paper}>
          <Grid container spacing={3}>
            <Grid item xs={12} align='center'>
              <Typography component='h1' variant='h5'>
                {user.firstName}'s Pantry
              </Typography>
            </Grid>
            <Grid item xs={12}>
              {pantry.data.map((ingredient, index) => {
                const formatName = _.startCase(_.toLower(ingredient.name));
                const formatQuantityType = _.startCase(
                  _.toLower(ingredient.quantityType)
                );

                return (
                  <PantryItem
                    key={index}
                    id={ingredient._id}
                    name={formatName}
                    userEmail={user.email}
                    quantityType={formatQuantityType}
                    quantity={ingredient.quantity}
                    handleDelete={handleDelete}
                    updatePantry={getPantryData}
                  />
                );
              })}
            </Grid>
            <Grid item xs={12}>
              {!isEmpty(error.errorMessage) && (
                <FormSubmitMessage submitMessage={error.errorMessage} />
              )}
            </Grid>
          </Grid>
        </div>
      </Card>
    </Container>
  );
}

export default Pantry;
