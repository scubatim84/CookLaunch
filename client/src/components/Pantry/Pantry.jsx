import React, {useState, useEffect} from 'react';
import {Redirect} from 'react-router-dom';
import _ from 'lodash';
import isEmpty from 'is-empty';
import {REQUEST_SUCCESS} from '../../actions/types';
import {getUserToken} from '../../actions/authActions';
import {
  getPantry,
  deleteIngredientFromPantry,
} from '../../actions/pantryActions';
import {useStylesForm} from '../../Styles';
import {themeMain} from '../../Theme';
import PantryAdd from './PantryAdd';
import PantryItem from './PantryItem';

import {
  Card,
  Container,
  CssBaseline,
  Grid,
  Typography,
} from '@material-ui/core';

function Pantry(props) {
  const classes = useStylesForm(themeMain);

  const [isLoggedIn, setLoggedIn] = useState(props.isLoggedIn);
  const [user, setUser] = useState({
    email: '',
    firstName: '',
    lastName: '',
  });
  const [pantry, setPantry] = useState({data: [{name: 'None'}]});

  const getPantryData = async () => {
    if (!isEmpty(user.email)) {
      const response = await getPantry(user.email);

      if (response.authResponseType === REQUEST_SUCCESS) {
        setPantry({data: response.authResponsePayload});
      }
    }
  };

  useEffect(() => {
    getPantryData(user.email);
  }, [user]);

  useEffect(() => {
    setLoggedIn(props.isLoggedIn);
  }, [props.isLoggedIn]);

  useEffect(() => {
    if (isLoggedIn) {
      const getUserPayload = async () => {
        const data = await getUserToken();
        const userPayload = await data.payload;

        setUser({
          email: userPayload.email,
          firstName: userPayload.firstName,
          lastName: userPayload.lastName,
        });

        setPantry({
          data: userPayload.pantry,
        });
      };

      getUserPayload();
    }
  }, [isLoggedIn]);

  const handleDelete = async (ingredientId) => {
    await deleteIngredientFromPantry(ingredientId);

    // Update pantry data to re-render
    await getPantryData();
  };

  return !isLoggedIn ? (
    <Redirect to='/login' />
  ) : (
    <Container component='main' maxWidth='md'>
      <Card className={classes.root}>
        <CssBaseline />
        <div className={classes.paper}>
          <Grid container spacing={3}>
            <Grid item xs={12} align='center'>
              <Typography component='h1' variant='h5'>
                {user.firstName}'s Pantry
              </Typography>
            </Grid>
            <Grid item xs={12}>
              {pantry.data.map((ingredient, index) => {
                const formatName = _.startCase(_.toLower(ingredient.name));
                const formatQuantityType = _.startCase(
                  _.toLower(ingredient.quantityType)
                );

                return (
                  <PantryItem
                    key={index}
                    id={ingredient._id}
                    name={formatName}
                    userEmail={user.email}
                    quantityType={formatQuantityType}
                    quantity={ingredient.quantity}
                    handleDelete={handleDelete}
                    updatePantry={getPantryData}
                  />
                );
              })}
            </Grid>
          </Grid>
        </div>
      </Card>
      <PantryAdd userEmail={user.email} updatePantry={getPantryData} />
    </Container>
  );
}

export default Pantry;
