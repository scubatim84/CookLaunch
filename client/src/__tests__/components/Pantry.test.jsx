import React from 'react';
import { shallow } from 'enzyme';
import Pantry from '../../components/Pantry';

describe('Pantry', () => {
  it('Renders component without crashing', () => {
    shallow(<Pantry />);
  });
});
