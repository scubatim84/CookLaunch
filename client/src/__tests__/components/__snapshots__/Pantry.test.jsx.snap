import React from 'react';
import {shallow} from 'enzyme';
import Pantry from '../../components/Pantry';

describe('Pantry', () => {
  it('Renders correctly', () => {
    const wrapper = shallow(<Pantry />);

    expect(wrapper).toMatchSnapshot();
  });
});
